server/svix-server/src/worker.rs

```diff
@@ -224,6 +227,93 @@ struct DispatchExtraIds<'a> {
     msg_uid: Option<&'a MessageUid>,
 }
 
+/// This is the [`Future`] returned by the [`DnsResolver`] [`tower::Service`]. It simply wraps
+/// the [`tokio::task::JoinHandle`] created when blocking on resolving the domain name and flattens
+/// the [`Result`] of a [`Result`].
+struct DnsResolverFuture<I: Iterator<Item = SocketAddr>> {
+    inner: tokio::task::JoinHandle<std::result::Result<I, std::io::Error>>,
+}
+
+impl<I: Iterator<Item = SocketAddr>> Future for DnsResolverFuture<I> {
+    type Output = std::result::Result<I, std::io::Error>;
+
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        Pin::new(&mut self.inner).poll(cx).map(|res| {
+            let res = res.map_err(|e| {
+                if e.is_cancelled() {
+                    std::io::Error::new(std::io::ErrorKind::Interrupted, e)
+                } else {
+                    panic!("DNS resolution background task failed: {:?}", e)
+                }
+            });
+            match res {
+                Ok(Ok(out)) => Ok(out),
+                Ok(Err(e)) => Err(e),
+                Err(e) => Err(e),
+            }
+        })
+    }
+}
+
+/// The [`DnsResolver`] is a simple struct implementing a [`tower::Service`] that lets it be used
+/// as a `hyper` DNS resolver for the default [`hyper::client::HttpConnector`]. This resolution does
+/// the exact same as `hyper`'s built-in [`
+#[derive(Clone)]
+struct DnsResolver {
+    allowed_private_subnets: Vec<ipnet::IpNet>,
+}
+
+impl tower::Service<hyper::client::connect::dns::Name> for DnsResolver {
+    type Response = Box<dyn Iterator<Item = SocketAddr> + Send>;
+    type Error = std::io::Error;
+    type Future = DnsResolverFuture<Self::Response>;
+
+    fn poll_ready(
+        &mut self,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<std::result::Result<(), Self::Error>> {
+        std::task::Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, name: hyper::client::connect::dns::Name) -> Self::Future {
+        let inner = tokio::task::spawn_blocking({
+            let allowed_private_subnets = self.allowed_private_subnets.clone();
+            move || {
+                let ips_res = (name.as_str(), 0).to_socket_addrs();
+
+                ips_res
+                    .map(move |ips| {
+                        ips.filter(move |socket| {
+                            let ip = socket.ip();
+
+                            let is_private = ip.is_loopback()
+                                || ip.is_multicast()
+                                || match ip {
+                                    IpAddr::V4(ip) => ip.is_private(),
+                                    IpAddr::V6(_) => false,
```

Comment:
IPV6 also have private subnets. You can look at the "is_global" code of the IPv6 variants of the underlying struct.
That's what you want really, just check if global.